timemachine/cpp/src/kernels/k_nonbonded.cuh

```diff
@@ -52,46 +52,52 @@ void __global__ k_check_rebuild_coords_and_box_gather(
     }
 }
 
-template <typename RealType>
+template <typename RealType, int COORDS_DIM, int PARAMS_DIM>
 void __global__ k_gather_coords_and_params(
     const int N,
     const unsigned int *__restrict__ idxs,
     const RealType *__restrict__ coords,
     const RealType *__restrict__ params,
     RealType *__restrict__ gathered_coords,
     RealType *__restrict__ gathered_params) {
-
+    static_assert(COORDS_DIM == 3);
+    static_assert(PARAMS_DIM == PARAMS_PER_ATOM);
     int idx = blockIdx.x * blockDim.x + threadIdx.x;
-    int stride = gridDim.y;
-    int stride_idx = blockIdx.y;
-
     if (idx >= N) {
         return;
     }
 
+    const unsigned int atom_idx = idxs[idx];
+
     // Coords have 3 dimensions, params have 4
-    if (stride_idx < 3) {
-        gathered_coords[idx * 3 + stride_idx] = coords[idxs[idx] * 3 + stride_idx];
+#pragma unroll COORDS_DIM
+    for (int i = 0; i < COORDS_DIM; i++) {
+        gathered_coords[idx * COORDS_DIM + i] = coords[atom_idx * COORDS_DIM + i];
+    }
+#pragma unroll PARAMS_DIM
+    for (int i = 0; i < PARAMS_DIM; i++) {
+        gathered_params[idx * PARAMS_DIM + i] = params[atom_idx * PARAMS_DIM + i];
     }
-    gathered_params[idx * stride + stride_idx] = params[idxs[idx] * stride + stride_idx];
 }
 
-template <typename T>
+template <typename T, int D>
 void __global__ k_scatter_accum(
     const int N,
     const unsigned int *__restrict__ unique_idxs, // NOTE: race condition possible if there are repeated indices
     const T *__restrict__ gathered_array,
     T *__restrict__ array) {
-
+    static_assert(D <= PARAMS_PER_ATOM);
```

Comment:
Can this assertion be moved to the call site? (This was previously a generic function; seems preferable to not unnecessarily specialize it to dealing with parameter arrays)

Reply:
In practice this is a specialized function, we strictly use it for `du_dx` and `params`. Would rather resolve this when we use it differently than trying to adding more duplicate code.

Reply:
That's fine, but just want to argue that even if this is the only way we use it now, keeping it generic has benefits for conceptual clarity too. E.g. it becomes more obvious that the implementation is independent of details like `PARAMS_PER_ATOM`.

Reply:
(to be clear, this is a non-blocking nit)

> We don't want to start unrolling loops that are super long.

Would it be more appropriate to use a threshold value like "longest loop we think it makes sense to unroll", like 10 or 100, rather than coupling to other code? I could imagine increasing `PARAMS_PER_ATOM` to a larger value where the unroll optimization no longer makes sense.

Reply:
https://github.com/proteneer/timemachine/pull/1124/commits/e37a5b30bf9d378fcd1f1683f1e26b724b08a9eb

Set to 5 to be safe, would want to verify the performance difference as we increase these values by more than one or two